utils/id_hash.go

```diff
@@ -0,0 +1,45 @@
+package utils
+
+import (
+	"bytes"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	u "github.com/ipfs/go-ipfs-util"
+	blake2b "golang.org/x/crypto/blake2b"
+)
+
+// ID signifies that its contents have been hashed from either a
+// peer.ID or a util.Key. This unifies the keyspace.
+type ID []byte
+
+// HashCache stores hashes of peer IDs.
+type HashCache map[peer.ID]ID
+
+// Less return true is id is less than other.
+func (id ID) Less(other ID) bool {
+	return bytes.Compare([]byte(id), []byte(other)) < 0
+}
+
+// XOR return xor of two byte arrays.
+func XOR(a, b ID) ID {
+	return ID(u.XOR(a, b))
+}
+
+// ConvertPeerID creates a DHT ID by hashing a Peer ID (Multihash).
+func (h HashCache) ConvertPeerID(id peer.ID) ID {
+	hash, ok := h[id]
+	if !ok {
+		hashBytes := blake2b.Sum256([]byte(id))
```

Comment:
ConvertKey takes a string, id is not.

I could use ConvertKey(string(id)), but that would give a different hash as []byte(id) and []byte(string(id)) are not same.
I could have the same function to hash both of these things. ConvertKey(interface{}), but since we don't want to cache cid hashes (they appear only once), having ConvertKey helps.